### ISSUE TITLE
[JSX] Do not convert html-escaped entities to their prettier forms

### ISSUE THREAD

**Issue by markguckian** [prettier/prettier#678]
Most HTML entity names & numbers are converted to their literal form.
Seems `&nbsp;` `&lt;` `&gt;` are the only exceptions? 
(`&nbsp;` escaping was fixed in #557)

Is this expected?

```js
<span>
  // not converted
  &nbsp; &#160;
  &lt; &#60;
  &gt; &#62;
  // converted
  &amp; &#38;
  &quot; &#34;
  &apos; &#39;
  &aacute; &#225;
</span>
```

turns into

```js
<span>
  // not converted
  &nbsp; &nbsp;
  &lt; &lt;
  &gt; &gt;
  // converted
  & &
  " "
  ' '
  á á
</span>;
```

https://jlongster.github.io/prettier/#%7B%22content%22%3A%22%3Cspan%3E%5Cn%20%20%2F%2F%20not%20converted%5Cn%20%20%26nbsp%3B%20%26%23160%3B%5Cn%20%20%26lt%3B%20%26%2360%3B%5Cn%20%20%26gt%3B%20%26%2362%3B%5Cn%20%20%2F%2F%20converted%5Cn%20%20%26amp%3B%20%26%2338%3B%5Cn%20%20%26quot%3B%20%26%2334%3B%5Cn%20%20%26apos%3B%20%26%2339%3B%5Cn%20%20%26aacute%3B%20%26%23225%3B%5Cn%3C%2Fspan%3E%5Cn%5Cn%22%2C%22options%22%3A%7B%22printWidth%22%3A80%2C%22tabWidth%22%3A2%2C%22singleQuote%22%3Afalse%2C%22trailingComma%22%3Afalse%2C%22bracketSpacing%22%3Atrue%2C%22doc%22%3Afalse%7D%7D

Thanks.

**Comment by vjeux** [comment]
@rattrayalex how hard would it be to keep the original way things are escaped like we are now doing for strings inside of JSX?

**Comment by rattrayalex** [comment]
I'm not sure offhand how hard it would be, but fwiw the answer to the question "is this expected?" is, from my perspective, "yes". My personal feelings is that HTML entities should be displayed in the prettiest possible form. The only ones for which that involves escaping are `&nbsp;`, `&lt;` and `&rt;`. Nbsp because it's invisible, and lt and rt because they conflict with JSX.

**Comment by rattrayalex** [comment]
@markguckian what are your thoughts on this? Are there reasons you would like to see, for example, `&quot;` or `&#34;` in your source code?

**Comment by vjeux** [comment]
The biggest thing that came up with the string encoding is that some people want to have the \u form and some want to see their emojis inline. We tried to encode nothing, and then encode everything and we found the best trade-off was to leave things as they were inputted.

For changing `&#34;` to `&quot;` I think is fine

**Comment by markguckian** [comment]
@rattrayalex I tend to use entity names in code for some obscure characters as they're easier to remember than finding the actual char e.g. `&ldquo;` `&mdash;` `&hellip;` etc.
My question concerned the expected behaviour and I'm ok with conversion being consistent across all entities (with exception of the 3 listed above).

BTW thanks very much for the hard work guys!

**Comment by rattrayalex** [comment]
@vjeux  – ah, gotcha. So JSX is currently inconsistent from the rest of prettier in that regard? 

@markguckian – thanks! those are good examples of characters that are probably better written out. 

Now that I'm refreshing on the context here, it might actually not be possible until https://github.com/babel/babel/pull/5256 lands (and it might not land at all), because what we really want to do is use the `raw` field, which is currently stripped out prematurely. So I think if users want this feature they will need to use the `flow` parser in the meantime. 

Here's the output for the flow parser:

```jsx
<span>
  // not converted
  &nbsp; &#160;
  &lt; &#60;
  &gt; &#62;
  // converted
  &amp; &#38;
  &quot; &#34;
  &apos; &#39;
  &aacute; &#225;
</span>;
```

**Comment by rattrayalex** [comment]
https://github.com/babel/babylon/pull/344 and https://github.com/babel/babel/pull/5256 have been approved – I am told they should be merged into Babel 7.0.

**Comment by rhengles** [comment]
https://github.com/babel/babel/pull/5256 is merged! 🎉

**Comment by oscar-b** [comment]
Another example is soft hyphen: `&shy;` which is transformed to an invisible character (at least in vscode) when using Baylon parser: https://prettier.github.io/prettier/#%7B%22content%22%3A%22const%20Terms%20%3D%20()%20%3D%3E%20(%5Cn%5Ct%3CPage%3E%5Cn%5Ct%5Ct%3Cheader%3E%5Cn%5Ct%5Ct%5Ct%3Ch1%3E%5Cn%5Ct%5Ct%5Ct%5CtDeltagar%26shy%3Bvillkor%5Cn%5Ct%5Ct%5Ct%3C%2Fh1%3E%5Cn%5Ct%5Ct%3C%2Fheader%3E%5Cn%20%20%5Ct%3C%2FPage%3E%5Cn%20%20)%22%2C%22options%22%3A%7B%22printWidth%22%3A80%2C%22tabWidth%22%3A2%2C%22singleQuote%22%3Atrue%2C%22trailingComma%22%3A%22none%22%2C%22bracketSpacing%22%3Atrue%2C%22jsxBracketSameLine%22%3Afalse%2C%22parser%22%3A%22babylon%22%2C%22semi%22%3Atrue%2C%22useTabs%22%3Atrue%2C%22doc%22%3Afalse%7D%7D

**Comment by rattrayalex** [comment]
@oscar-b can you open a separate issue for that?

**Comment by rattrayalex** [comment]
@markguckian I'm not sure about the status of this issue... should it be closed?

**Comment by markguckian** [comment]
@rattrayalex don't think so - the HTML entities are still converted (using the babylon parser in Prettier 1.2.2)
(click on the link in my original post above to verify)

**Comment by rattrayalex** [comment]
Ah, sorry. I always lose track because the lede is a bit buried in this thread. 

It seems the desired behavior (which, fwiw, I mildly disagree with) is to leave *all* HTML-escaped characters as input – for example, `&amp;` and `&#38;` remain as they are, instead of being converted to `&`. 

Does that sound correct? @yamafaktory is that the behavior of https://github.com/prettier/prettier/pull/1165 ?

**Comment by yamafaktory** [comment]
@rattrayalex Yes it is!

**Comment by MrEfrem** [comment]
I also have the problem with JSX:
`<strong>&#123;&#123;{next.name}&#125;&#125;</strong>`
have converting to:
`<strong>{{{next.name}}}</strong>`
what causes a parcer error

**Comment by lydell** [comment]
@MrEfrem That has been fixed in `master` and should be available in the next release!

**Comment by yamafaktory** [comment]
@MrEfrem this should have been fixed by #1165 🤞.

**Comment by vjeux** [comment]
This was fixed by #1165
